vitejs/vite
# fix(build): decode urls in CSS files (fix #15109)

### Description


- Call `decodeURI` in `urlReplacer` in the CSS plugin to decode CSS urls

### Additional context

Previously, `url()` calls in CSS were not decoded at build time. Web browsers *do* decode this call, so any `url()` call with an encoded string as a parameter would not resolve at build time, resulting in the following error:

```js
`\n${url} referenced in ${id} didn't resolve at build time, it will remain unchanged to be resolved at runtime`
```

This is important for some font files, which contain special characters in their names for variable weight/width. This PR fixes this by calling `decodeURI` at the top of `urlReplacer`, converting characters like `%5B` and `%5D` into their actual characters `[` and `]` respectively. 

More information on this bug can be found here: #15109 

#### QA steps

1. Clone this branch and go to `packages/vite`, and run `pnpm run dev`
2. Create a new Vite project in a different directory (I tested with React + TypeScript)
3. Follow the instructions in [`CONTRIBUTING.MD`](https://github.com/vitejs/vite/blob/0654d1b52448db4d7a9b69aee6aad9e015481452/CONTRIBUTING.md#testing-vite-against-external-packages) to link this branch to the new Vite project
4. Create a subfolder `fonts` and place a font file inside, ensuring the name contains `[` and `]`
5. In `index.css`, define the font face, but replace `[` and `]` with `%5B` and `%5D` respectively in the `src` property. Example:
```css
@font-face {
  font-family: 'Ubuntu variable';
  font-stretch: 100%; /*  min and max value for the width axis, expressed as percentage */
  font-style: normal;
  font-weight: 100 800; /*  min and max value for the weight axis */
  src: url('/fonts/f1ea362b-Ubuntu%5Bwdth,wght%5D-latin-v0.896a.woff2')
    format('woff2-variations');
}
```
6. Run `pnpm run build`, and ensure the font file is resolved.

Note: The font file will resolve with underscores replacing special characters in the build output directory - this behaviour can be mitigated by setting `build.rollupOptions.output.sanitizeName` to `false`.



---

### What is the purpose of this pull request? 


#### Fixes

Fixes #15109 

### Before submitting the PR, please make sure you do the following

- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md), especially the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines).
- [ ] ~~Update the corresponding documentation if needed.~~
- [x] Ideally, include relevant tests that fail without this PR but pass with it.
